MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,2059 @@
+        if (verbose)
+          printf("   merge worked!\n");
+
+        if (vpeff)
+          vpeff->kind(VertexerPairEff::merge);
+
+        vertices->erase(v[1]);
+        *v[0] = reco::Vertex(new_vertices[0]); // ok to use v[0] after the erase(v[1]) because v[0] is by construction before v[1]
+      }
+      else {
+        if (verbose)
+          printf("   merge didn't work, trying arbitration refits\n");
+        refit = true;
+      }
+
+      if (verbose)
+        printf("   vertices size is now %lu\n", vertices->size());
+    }
+
+    if (refit) {
+      bool erase[2] = { false };
+      reco::Vertex vsave[2] = { *v[0], *v[1] };
+
+      for (int i = 0; i < 2; ++i) {
+        if (tracks_to_remove_in_refit[i].empty())
+          continue;
+
+        if (verbose) {
+          printf("   refit vertex%i %lu with these tracks:", i, ivtx[i]);
+          print_track_set(tracks[i]);
+          printf("   but skip these:");
+          print_track_set(tracks_to_remove_in_refit[i]);
+          printf("\n");
+        }
+
+        std::vector<reco::TransientTrack> ttks;
+        for (auto tk : tracks[i])
+          if (tracks_to_remove_in_refit[i].count(tk) == 0)
+            ttks.push_back(seed_tracks[seed_track_ref_map[tk]]);
+
+        reco::VertexCollection new_vertices;
+        for (const TransientVertex& tv : kv_reco_dropin(ttks))
+          new_vertices.push_back(reco::Vertex(tv));
+        if (verbose) {
+          printf("      got %lu new vertices out of the av fit for v%i\n", new_vertices.size(), i);
+          printf("      these track sets:");
+          for (const auto& nv : new_vertices) {
+            printf(" (");
+            print_track_set(nv);
+            printf(" ),");
+          }
+          printf("\n");
+        }
+        if (new_vertices.size() == 1)
+          * v[i] = new_vertices[0];
+        else
+          erase[i] = true;
+      }
+
+      if (vpeff && (erase[0] || erase[1]))
+        vpeff->kind(VertexerPairEff::erase);
+
+      if (erase[1]) vertices->erase(v[1]);
+      if (erase[0]) vertices->erase(v[0]);
+
+      if (verbose)
+        printf("      vertices size is now %lu\n", vertices->size());
+    }
+
+    // If we changed the vertices at all, start loop over completely.
+    if (duplicate || merge || refit) {
+      v[0] = vertices->begin() - 1;  // -1 because about to ++sv
+      ++n_resets;
+      if (verbose) printf("   resetting from vertices %lu and %lu. # of resets: %i\n", ivtx[0], ivtx[1], n_resets);
+
+      //if (n_resets == 3000)
+      //  throw "I'm dumb";
+    }
+  }
+
+  if (verbose)
+    printf("n_resets: %i  n_onetracks: %i  n_noshare_vertices: %lu\n", n_resets, n_onetracks, vertices->size());
+  if (histos) {
+    h_n_resets->Fill(n_resets);
+    h_n_onetracks->Fill(n_onetracks);
+    h_n_noshare_vertices->Fill(vertices->size());
+  }
+
+  if (histos_noshare || verbose || do_track_refinement) {

```

Comment:
I propose that we now remove the `|| verbose` part here. A verbose mode is great for debugging, but shouldn't be used to do extra steps when it can be avoided (which I know is something that was in place before you made your additions!). 

Let me know if you agree!

Reply:
Similarly: do we want it when `histos_noshare` is on, or do we really only want it when the track refinement is being used?

Reply:
Thanks, I think it makes sense to me now. Adjusted in https://github.com/DisplacedVertices/cmssw-usercode/commit/70f6416b74dab704a3ab83a15671103f2daca491, but let me know if I didn't quite capture the intention.